vespalib/src/vespa/vespalib/util/spin_lock.h

```diff
@@ -0,0 +1,33 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <atomic>
+
+namespace vespalib {
+
+/**
+ * A spin-lock implementation that favors uncontended performance.
+ * Some measures are taken to reduce the impact of threads waiting to
+ * get the lock since this will not affect the fast-path of obtaining
+ * the lock immediately.
+ *
+ * This implementation satisfies the BasicLockable requirements,
+ * making it work with things like std::lock_guard.
+ **/
+class SpinLock {
+private:
+    std::atomic<bool> _lock;
+public:
+    SpinLock() noexcept : _lock(false) {
+        static_assert(std::atomic<bool>::is_always_lock_free);
+    }
+    void lock() noexcept {
+        while (__builtin_expect(_lock.exchange(true, std::memory_order_acquire), false)) {
+            while (_lock.load(std::memory_order_relaxed)) {
+                __builtin_ia32_pause();                
```

Comment:
`_mm_pause` may be more portable, but I believe it requires `xmmintrin.h` so if used maybe it should be wrapped away in its own translation unit to avoid header pollution.

Reply:
I could not figure out the subtly differences between _mm_pause and __builtin_ia32_pause across compilers, but concluded that if __builtin_ia32_pause is supported, it generates the appropriate code. I will let @toregge sink into this when he tries to compile Vespa on other platforms.